### PR TITLE
fixes a bug where OData Count path items would be missing from the description

### LINKED ISSUE #129: Count path segments are missing after translation
[OData supports counts segements](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_AddressingtheCountofaCollection) by conventions. e.g. GET /users/$count

All these segments are missing from the resulting OpenAPI description.

### PR DESCRIPTION
**Pull request by baywet**
- pattern matching refactoring
- adds launch configuration for update docs tool
- adds update docs to solution file so debug works properly
- code linting
- code linting
- fixes #129 a bug where odata count paths would be missing from the description

## TODO
- [x] https://github.com/oasis-tcs/odata-vocabularies/blob/main/vocabularies/Org.OData.Capabilities.V1.xml#L210

### CODE REVIEW

File: src/Microsoft.OpenApi.OData.Reader/Common/Constants.cs [hunk 1]
```diff
@@ -79,5 +79,9 @@ internal static class Constants
         /// extension for discriminator value support
         /// </summary>
         public static string xMsDiscriminatorValue = "x-ms-discriminator-value";
+        /// <summary>
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/Common/Constants.cs:82]: insert a white-line

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataDollarCountSegment.cs [hunk 2]
```diff
@@ -12,6 +12,10 @@ namespace Microsoft.OpenApi.OData.Edm
     /// </summary>
     public class ODataDollarCountSegment : ODataSegment
     {
+        /// <summary>
+        /// Get the static instance of $count segment.
+        /// </summary>
+        internal static ODataDollarCountSegment Instance = new();
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataDollarCountSegment.cs:18]: insert a white-line after

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPath.cs [hunk 3]
```diff
@@ -146,8 +146,8 @@ public string GetPathItemName(OpenApiConvertSettings settings)
 
             // From Open API spec, parameter name is case sensitive, so don't use the IgnoreCase HashSet.
             // HashSet<string> parameters = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
-            HashSet<string> parameters = new HashSet<string>();
-            StringBuilder sb = new StringBuilder();
+            HashSet<string> parameters = new();
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPath.cs:149]: If possible, let's keep it unchanged?

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 4]
```diff
@@ -360,6 +377,19 @@ private void CreateRefPath(ODataPath currentPath)
             newPath.Push(ODataRefSegment.Instance); // $ref
             AppendPath(newPath);
         }
+        /// <summary>
+        /// Create $count paths.
+        /// </summary>
+        /// <param name="currentPath">The current OData path.</param>
+        /// <param name="convertSettings">The settings for the current conversion.</param>
+        private void CreateCountPath(ODataPath currentPath, OpenApiConvertSettings convertSettings) {
+            if(currentPath == null) throw new ArgumentNullException(nameof(currentPath));
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:386]: make a better format

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 5]
```diff
@@ -360,6 +377,19 @@ private void CreateRefPath(ODataPath currentPath)
             newPath.Push(ODataRefSegment.Instance); // $ref
             AppendPath(newPath);
         }
+        /// <summary>
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:380]: insert a white-line

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 4 above)
Comment (xuzhg) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:386]: for others

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs [hunk 6]
```diff
@@ -169,20 +169,26 @@ private void AppendPath(ODataPath path)
...
 
             IEdmEntitySet entitySet = navigationSource as IEdmEntitySet;
             IEdmEntityType entityType = navigationSource.EntityType();
+            CountRestrictionsType count = null;
 
-            // for entity set, create a path with key
+            // for entity set, create a path with key and a $count path
             if (entitySet != null)
             {
+                count = _model.GetRecord<CountRestrictionsType>(entitySet, CapabilitiesConstants.CountRestrictions);
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:188]: shall we add a config in the settings to allow/disallow the $count segment

File: src/Microsoft.OpenApi.OData.Reader/Edm/RecordExpressionExtensions.cs [hunk 7]
```diff
@@ -29,20 +29,10 @@ internal static class RecordExpressionExtensions
...
-                    IEdmIntegerConstantExpression value = property.Value as IEdmIntegerConstantExpression;
-                    if (value != null)
-                    {
-                        return value.Value;
-                    }
-                }
-            }
-
-            return null;
+            return (record.Properties?.FirstOrDefault(e => e.Name == propertyName) is IEdmPropertyConstructor property &&
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/Edm/RecordExpressionExtensions.cs:32]: all of these code change are not related to the $count path feature. 
Maybe better to have a separate PR for clarification, better understanding.

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPath.cs (hunk 3 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPath.cs:149]: what's the harm in using the newer syntax that saves code writing and removes squiggles when we're editing the code?

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 4 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:386]: I'm assuming that you're referring to the curly brace not being on the next line. Fixed.

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 6 above)
Comment (xuzhg) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:188]: Ok, I see that. IncludeDollarCountPathSegments.
But, why don't you check it then call "GetRecord". If it's not allowed, we don't need to call "GetRecord".

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 5 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:380]: fixed.

File: test/Microsoft.OpenAPI.OData.Reader.Tests/Resources/Basic.OpenApi.json [hunk 8]
```diff
@@ -293,6 +293,27 @@
         "x-ms-docs-operation-type": "operation"
       }
     },
+    "/City/$count": {
```
Comment (xuzhg) [inline test/Microsoft.OpenAPI.OData.Reader.Tests/Resources/Basic.OpenApi.json:296]: City looks like a singleton. 
It's not allowed to have $count after a single navigation property or singleton.

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 6 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:188]: the setting is already read in the `CreateCountPath` method

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataDollarCountSegment.cs (hunk 2 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataDollarCountSegment.cs:18]: fixed

File: src/Microsoft.OpenApi.OData.Reader/Common/Constants.cs (hunk 1 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Common/Constants.cs:82]: fixed

File: src/Microsoft.OpenApi.OData.Reader/Edm/RecordExpressionExtensions.cs (hunk 7 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Edm/RecordExpressionExtensions.cs:32]: I'll try to keep that in mind for the next pull requests. I usually try to fix warnings/suggestions on the files I touch, but not the ones I don't touch. As this is my first actual PR in this codebase, I was exploring the code to understand how things work and I made a few modifications at other places. Since the work was already done, and it's only syntax change that doesn't impact the behavior, I suggest we make a one time exception. Thoughts?

File: test/Microsoft.OpenAPI.OData.Reader.Tests/Resources/Basic.OpenApi.json (hunk 8 above)
Comment (baywet) [reply test/Microsoft.OpenAPI.OData.Reader.Tests/Resources/Basic.OpenApi.json:296]: city is actually also an entity set, arguably it should be names cities but I didn't want to impact the source file.
https://github.com/microsoft/OpenAPI.NET.OData/blob/13427946decd6bbe1ba5e7b52604d32698f455d1/test/Microsoft.OpenAPI.OData.Reader.Tests/Common/EdmModelHelper.cs#L273

**Comment by xuzhg** [comment]
I'd suggest some "rules" about "coding style":
1) leave new T() unchanged to use new(), until we have whole separate PR to clean all.
2) Insert white-line between paragraph.
3) If (....) xxxx =>
     if (....)
      {
          ...
      }
      
      even it's only one line code.

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 6 above)
Comment (xuzhg) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:188]: if (setting.IncludeDollarCountPathSegments)
{
     call CreateCountPath(...)
}

in the CreateCountPath()
{
      Get CountRestirction()
      then do something based on the CountRestriciton?
}

File: src/Microsoft.OpenApi.OData.Reader/OpenApiConvertSettings.cs [hunk 9]
```diff
@@ -183,6 +183,11 @@ public string PathPrefix
         /// </summary>
         public IODataPathProvider PathProvider { get; set; }
 
+        /// <summary>
+        /// Gets/sets a value indicating whether or not add OData $count segments in the description for collections.
+        /// </summary>
+        public bool IncludeDollarCountPathSegments { get; set; } = true;
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/OpenApiConvertSettings.cs:189]: IncludeDollarCountPathSegments -> EnableDollarCountPath

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 6 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:188]: why read the setting in the before the call to CreateCountPath? It's only going to lead to code/logic duplication.

File: src/Microsoft.OpenApi.OData.Reader/OpenApiConvertSettings.cs (hunk 9 above)
Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/OpenApiConvertSettings.cs:189]: renamed

**Comment by baywet** [comment]
@xuzhg I'm going to push back on those rules until we have a comprehensive editorconfig + linter to enforce them (and resolve them automatically). It's a poor use of our time to enforce those things manually. And I do agree on the fact that the bulk of changes + configuration should happen in a separate PR.

File: Microsoft.OpenApi.OData.sln [hunk 10]
```diff
@@ -16,6 +16,10 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution
 		.editorconfig = .editorconfig
 	EndProjectSection
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "tool", "tool", "{DE8F8E75-A119-4CF3-AFDD-4132B55DAE76}"
```
Comment (xuzhg) [inline Microsoft.OpenApi.OData.sln:19]: I remember I have a separate solution for updateDocs.

Do you think it's convenient to have it into the main solution? Because it's only a helper tooling.

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPath.cs (hunk 3 above)
Comment (xuzhg) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPath.cs:149]: I don't think it can save a lot of time. It's the same as:  

`var parameters = new HashSet<string>();`

Actually, it's no harm, but for readability.

File: src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs (hunk 4 above)
Comment (xuzhg) [reply src/Microsoft.OpenApi.OData.Reader/Edm/ODataPathProvider.cs:388]: if(currentPath == null)
{
      throw new ArgumentNullException(nameof(currentPath));
 }
 
 and, maybe to use Error.ArgumentNull(...) method?

File: src/Microsoft.OpenApi.OData.Reader/Edm/RecordExpressionExtensions.cs (hunk 7 above)
Comment (xuzhg) [reply src/Microsoft.OpenApi.OData.Reader/Edm/RecordExpressionExtensions.cs:32]: Oh, Ok

File: src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs [hunk 11]
```diff
@@ -25,15 +25,16 @@ internal class DollarCountGetOperationHandler : OperationHandler
         /// this segment could be "entity set", "Collection property", "Composable function whose return is collection",etc.
         /// </summary>
         internal ODataSegment LastSecondSegment { get; set; }
-
+        private const int SecondLastSegmentIndex = 2;
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs:28]: insert a new line before and after

File: src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs [hunk 12]
```diff
@@ -54,10 +55,12 @@ protected override void SetBasicInfo(OpenApiOperation operation)
         /// <inheritdoc/>
         protected override void SetResponses(OpenApiOperation operation)
         {
-            OpenApiSchema schema = new OpenApiSchema
-            {
-                Type = "integer",
-                Format = "int32"
+            OpenApiSchema schema = new()
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/Operation/DollarCountGetOperationHandler.cs:58]: align the codes

File: src/Microsoft.OpenApi.OData.Reader/PathItem/NavigationPropertyPathItemHandler.cs [hunk 13]
```diff
@@ -46,8 +45,8 @@ internal class NavigationPropertyPathItemHandler : PathItemHandler
         /// </summary>
         protected bool LastSegmentIsRefSegment { get; private set; }
 
-        /// <inheritdoc/>
-        protected override void SetOperations(OpenApiPathItem item)
+		/// <inheritdoc/>
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/PathItem/NavigationPropertyPathItemHandler.cs:48]: align

File: test/Microsoft.OpenAPI.OData.Reader.Tests/Edm/ODataPathProviderTests.cs [hunk 14]
```diff
@@ -81,6 +81,24 @@ public void GetPathsForInheritanceModelWithoutDerivedTypesConstraintReturnsMore(
...
+            Assert.Equal(4, paths.Count());
+        }
+
+        [Fact]
+        public void GetPathsDoesntReturnPathsForCountWhenDisabled()
+        {
+            // Arrange
+            IEdmModel model = GetInheritanceModel(string.Empty);
+            ODataPathProvider provider = new ODataPathProvider();
+            var settings = new OpenApiConvertSettings {
```
Comment (xuzhg) [inline test/Microsoft.OpenAPI.OData.Reader.Tests/Edm/ODataPathProviderTests.cs:95]: { in a separate linke

File: test/Microsoft.OpenAPI.OData.Reader.Tests/Resources/Basic.OpenApi.json (hunk 8 above)
Comment (xuzhg) [reply test/Microsoft.OpenAPI.OData.Reader.Tests/Resources/Basic.OpenApi.json:296]: Oh, ok. My error

**Review by xuzhg — APPROVED** [review]

**Comment by xuzhg** [comment]
@baywet I don't like the auto-merge setting.  Should turn it off.

Besides, we should merge the changes squash. It's better for one feature/one fix into one commit.

**Comment by baywet** [comment]
Squash: sure we can disable merge and rebase in the branch policy if you prefer linear history. Although it leads to more conflicts.

Auto-merge: what's the issue with it? It saves time to everyone.

File: Microsoft.OpenApi.OData.sln (hunk 10 above)
Comment (baywet) [reply Microsoft.OpenApi.OData.sln:19]: Having it in the main solution makes f5 debug work for VS code

Comment (baywet) [reply Microsoft.OpenApi.OData.sln:19]: Also enables static analysis and more for that project with CI

**Comment by baywet** [comment]
And I can't reply to your var comment for some reason. But I saw the style was not to use var here, I usually go with the var syntax to avoid duplicating the type name
